windows/windows_mphal.c

```diff
@@ -0,0 +1,175 @@
+    mode &= ~ENABLE_PROCESSED_INPUT;
+    SetConsoleMode(std_in, mode);
+}
+
+void mp_hal_stdio_mode_orig(void) {
+    assert(std_in);
+    SetConsoleMode(std_in, orig_mode);
+}
+
+void mp_hal_set_interrupt_char(char c) {
+    assert(std_in);
+    if (c == CHAR_CTRL_C) {
+        DWORD mode;
+        GetConsoleMode(std_in, &mode);
+        mode |= ENABLE_PROCESSED_INPUT;
+        SetConsoleMode(std_in, mode);
+    } else {
+        DWORD mode;
+        GetConsoleMode(std_in, &mode);
+        mode &= ~ENABLE_PROCESSED_INPUT;
+        SetConsoleMode(std_in, mode);
+    }
+}
+
+void mp_hal_move_cursor_back(uint pos) {
+    assert(con_out);
+    CONSOLE_SCREEN_BUFFER_INFO info;
+    GetConsoleScreenBufferInfo(con_out, &info);
+    info.dwCursorPosition.X -= (short)pos;
+    if (info.dwCursorPosition.X < 0) {
+        info.dwCursorPosition.X = 0;
+    }
+    SetConsoleCursorPosition(con_out, info.dwCursorPosition);
+}
+
+void mp_hal_erase_line_from_cursor() {
+    assert(con_out);
+    CONSOLE_SCREEN_BUFFER_INFO info;
+    GetConsoleScreenBufferInfo(con_out, &info);
+    const short len = info.dwSize.X - info.dwCursorPosition.X;
+    DWORD written;
+    FillConsoleOutputCharacter(con_out, ' ', len, info.dwCursorPosition, &written);
+    FillConsoleOutputAttribute(con_out, info.wAttributes, len, info.dwCursorPosition, &written);
+}
+
+typedef struct item_t {
+    WORD vkey;
+    const char *seq;
+} item_t;
+
+// map virtual key codes to VT100 escape sequences
+STATIC item_t keyCodeMap[] = {
+    {VK_UP, "[A"},
+    {VK_DOWN, "[B"},
+    {VK_RIGHT, "[C"},
+    {VK_LEFT, "[D"},
+    {VK_HOME, "[H"},
+    {VK_END, "[F"},
+    {VK_DELETE, "[3~"},
+    {0, ""} //sentinel
+};
+
+STATIC const char *cur_esc_seq = NULL;
+
+STATIC int esc_seq_process_vk(int vk) {
+    for (item_t *p = keyCodeMap; p->vkey != 0; ++p) {
+        if (p->vkey == vk) {
+            cur_esc_seq = p->seq;
+            return 27; // ESC, start of escape sequence
+        }
+    }
+    return 0; // nothing found
+}
+
+STATIC int esc_seq_chr() {
+    if (cur_esc_seq) {
+        const char c = *cur_esc_seq++;
+        if (c) {
+            return c;
+        }
+        cur_esc_seq = NULL;
+    }
+    return 0;
+}
+
+int mp_hal_stdin_rx_chr(void) {
+    // currently processing escape seq?
+    const int ret = esc_seq_chr();
+    if (ret) {
+        return ret;
+    }
+
+    // poll until key which we handle is pressed
+    DWORD num_read;
+    INPUT_RECORD rec;
+    for (;;) {
+      if (!ReadConsoleInput(std_in, &rec, 1, &num_read) || !num_read) {
+          return CHAR_CTRL_C; // EOF, ctrl-D
+      }
+      if (rec.EventType != KEY_EVENT || !rec.Event.KeyEvent.bKeyDown) { // only want key down events
+          continue;
+      }
+      const char c = rec.Event.KeyEvent.uChar.AsciiChar;
+      if (c) { // plain ascii char, return it
+          return c;
+      }
+      const int ret = esc_seq_process_vk(rec.Event.KeyEvent.wVirtualKeyCode);
+      if (ret) {
+          DEBUG_printf("vk");
```

Comment:
It can since I have DEBUG_printf print to the debugger instead of the commandline. But yes it's obviously an oversight of mine, fixed.